lib/rules/no-dupe-disjunctions.ts

```diff
@@ -22,106 +630,180 @@ export default createRule("no-dupe-disjunctions", {
+            overlap:
+                "Unexpected overlap. This alternative overlaps with '{{others}}'. The overlap is '{{expr}}'.{{cap}}{{exp}}",
         },
         type: "suggestion", // "problem",
     },
     create(context) {
-        const disallowNeverMatch = Boolean(
-            context.options[0]?.disallowNeverMatch,
-        )
+        const alwaysReportExponentialBacktracking =
+            context.options[0]?.alwaysReportExponentialBacktracking ?? true
+        const report: ReportOption =
+            context.options[0]?.report ?? ReportOption.trivial
 
         /**
-         * Check has after pattern
+         * Create visitor
          */
-        function hasAfterPattern(
-            node: Group | CapturingGroup | Pattern | LookaroundAssertion,
-        ): boolean {
-            if (node.type === "Assertion") {
-                return false
-            }
-            if (node.type === "Pattern") {
-                return false
-            }
-            let target: Group | CapturingGroup | Quantifier = node
-            let parent = target.parent
-            while (parent) {
-                if (parent.type === "Alternative") {
-                    const index = parent.elements.indexOf(target)
-                    if (index < parent.elements.length - 1) {
-                        return true
-                    }
-                    return hasAfterPattern(parent.parent)
-                }
-                if (parent.type === "Quantifier") {
-                    target = parent
-                    parent = target.parent
-                    continue
+        function createVisitor(
+            regexpContext: RegExpContext,
+        ): RegExpVisitor.Handlers {
+            const {
+                patternAst,
+                flagsString,
+                flags,
+                node,
+                getRegexpLocation,
+            } = regexpContext
+
+            const parser = JS.Parser.fromAst({
+                pattern: patternAst,
+                flags: new RegExpParser().parseFlags(
+                    [
+                        ...new Set(
+                            (flagsString || "").replace(/[^gimsuy]/g, ""),
+                        ),
+                    ].join(""),
+                ),
+            })
+
+            /** Verify group node */
+            function verify(parentNode: ParentNode) {
+                // report all if the we report exp backtracking
+                const nodeReport =
+                    alwaysReportExponentialBacktracking && isStared(parentNode)
+                        ? ReportOption.all
+                        : report
+
+                const hasNothingAfter = hasNothingAfterNode(parentNode)
+
+                const rawResults = findDuplication(
+                    parentNode.alternatives,
+                    regexpContext,
+                    {
+                        fastAst: true,
```

Comment:
Should we change it to `false` to report many warnings?

```suggestion
                        fastAst: false,
```

Reply:
Yes. My bad.

Reply:
I re-enabled the AST approach and added a test case the NFA approach definitely can't detect.